### ISSUE TITLE
[SURE-6510] Kubectl wait does not timeout correctly through Rancher API.

### ISSUE THREAD

**Issue by ssosna** [rancher/rancher#27244]
Internal reference: [SURE-6510](https://jira.suse.com/browse/SURE-6510)

**What kind of request is this (question/bug/enhancement/feature request):**
Bug

When using kubectl wait --timeout=Xs command it does not timeout correctly after X seconds when accessing the K8s API through Rancher. When pointed directly to EKS API it works as expected.
Wait does end correctly when the condition which we wait for is met.

**Steps to reproduce (least amount of steps as possible):**
1) create a deployment that will never be ready (pod with resource requirements greater than a single node or a pod that always fails readiness probe)
2) use `kubectl wait --timeout=Xs` on that deployment and verify if it quits after X seconds

Example:
_$ ./kubectl describe pod/mysql-7cc9679b8f-vfwpw
Name:               mysql-7cc9679b8f-vfwpw
Namespace:          dev
Priority:           0
PriorityClassName:  <none>
Node:               <none>
Labels:             name=mysql
                    pod-template-hash=7cc9679b8f
Annotations:        kubernetes.io/psp: eks.privileged
Status:             Pending
IP:
Controlled By:      ReplicaSet/mysql-7cc9679b8f
Containers:
  mysql:
    Image:      mysql:5.6
    Port:       3306/TCP
    Host Port:  0/TCP
    Limits:
      memory:  200000Mi
    Requests:
      cpu:     215m
      memory:  150000Mi
Conditions:
  Type           Status
  PodScheduled   False
QoS Class:       Burstable
Node-Selectors:  <none>_

Run:
`kubectl wait --for=condition=Ready pod/mysql-7cc9679b8f-vfwpw --timeout=5s`
**Result:**
The kubectl will wait forever without any errors or messages.

**Environment information**
- Rancher -  v2.4.3
- User Interface - v2.4.22
- Helm - v2.16.5-rancher1
- Machine - v0.15.0-rancher36
- Kubectl vesions: 1.11.10, 1.13.12, 1.18.3, 1.17.6, 1.15.9

**Comment by moio** [comment]
Reopening, this was reported to still happen.

**Comment by moio** [comment]
I cannot reproduce this trivially, eg.

```
# create an ngnix deployment with wrong image name
curl -L https://k8s.io/examples/application/deployment.yaml | sed 's/nginx:1.14.2/ngnix-TESTTHISDOESNTEXIST/g' | kubectl apply -f -

# try waiting
kubectl wait deployment --for=condition=Ready --timeout=3s -n default nginx-deployment
# ...terminates as expected

export KUBECONFIG=/path/to/kubeconfig/from/rancher/kubeconfig.yaml

# try waiting
kubectl wait deployment --for=condition=Ready --timeout=3s -n default nginx-deployment
# ...terminates as expected
```

If anybody knows how to reproduce this, it will be super helpful. TIA

**Comment by betweenclouds** [comment]
@moio I opened a case with suse about this (00669777) there are the logs and tests I did.
Please say, if I can you provide you with more information.

**Comment by skanakal** [comment]
@moio I can reproduce this problem on rancher installed on rke2 and rke clusters... it is working fine when rancher on k3s. 

Rancher Version: 2.7.5
kubectl version 1.26.0, 1.25, 1.24 
Server Version: v1.24.14+rke2r1

I followed the steps mentioned in the above example: 

```
root@master-0:~# kubectl get pod oversized-pod-5f8bb6887c-qwl7b -o json | jq -r .status 
{
  "conditions": [
    {
      "lastProbeTime": null,
      "lastTransitionTime": "2023-07-12T05:07:31Z",
      "message": "0/1 nodes are available: 1 Insufficient cpu, 1 Insufficient memory. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.",
      "reason": "Unschedulable",
      "status": "False",
      "type": "PodScheduled"
    }
  ],
  "phase": "Pending",
  "qosClass": "Guaranteed"
}
root@master-0:~# 
```
while going thru rancher proxy, every time I see a 30sec delay(the wait is at kubectl watch condition) for never met conditions... 

```
root@master-0:~# export KUBECONFIG=ACE_kubeconfig.yaml
root@master-0:~# 
root@master-0:~# env | grep KUBECONFIG
KUBECONFIG=ACE_kubeconfig.yaml
root@master-0:~# 
root@master-0:~# time kubectl wait --for=condition=Ready pod/oversized-pod-5f8bb6887c-qwl7b --timeout=3s
error: timed out waiting for the condition on pods/oversized-pod-5f8bb6887c-qwl7b
real    0m30.085s                                                                            <<<<<<<<<<-----------
user    0m0.046s
sys    0m0.033s
root@master-0:~# 
```

verbose output look like this: 
`round_trippers.go:553] GET https://newrancherservers.example.com/k8s/clusters/local/api/v1/namespaces/default/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dwordpress-86cf5f4d55-v6lld&resourceVersion=709159&timeoutSeconds=455&watch=true 200 OK in 30000 milliseconds`

While the requests are going through the API server, the requests that were never fulfilled timed out exactly at the time mentioned in the command.

```
root@master-0:~# 
root@master-0:~# export KUBECONFIG=/etc/rancher/rke2/rke2.yaml
root@master-0:~# 
root@master-0:~# env | grep KUBECONFIG
KUBECONFIG=/etc/rancher/rke2/rke2.yaml
root@master-0:~# 
root@master-0:~# 
root@master-0:~# time kubectl wait --for=condition=Ready pod/oversized-pod-5f8bb6887c-qwl7b --timeout=3s
error: timed out waiting for the condition on pods/oversized-pod-5f8bb6887c-qwl7b
real    0m3.044s                                                                            <<<<<<<<<<-----------
user    0m0.039s
sys    0m0.016s
root@master-0:~#

**Comment by rmweir** [comment]
I've been looking at this for a while now. This is to document what we've tried and what we've look at it:

TLDR; still unsure however we have found that setting `use-http2` to "false" in the `nginx-ingress-controller` configmap resolves the issue. This is just a workaround and admittedly not an ideal one.

Used remote-debugger for rancher and debugger in kubectl to look for clues. Request is being cancelled in kubectl but then it is stuck waiting for informer to close. Kubectl client is stuck waiting for headers. 

Kubectl client is stuck here https://github.com/golang/net/blob/master/http2/frame.go#L237. 

When proxying request with k3s usually an io.eof error is returned to indicate the reader can close. For the setups with the issue this is never sent during read. Read hangs until the timeout set in ingress closes it. Tried to replace reader with one that writes io.eof for the request but causes an error in the client (kubectl). 

Nginx uses http2 by default. Setting `use-http2` to “false” in the nginx-ingress-controller configmap resolves the issue. 

Tried using multiple version of the nginx-ingress-controller. Tried upgrading and down grading. Tried building custom image with newest version + rancher changes. Tried nic without ranchers changes. None of these resolved the issue. 

Tried changing multiple settings in both the annotations of the nic and the nic configmap. Some of which are described here http://nginx.org/en/docs/http/ngx_http_proxy_module.html. Includes the following: 
`proxy-buffering` set to off 

`proxy-method` set to method 

`proxy-request-buffering` set to off 

Along with using different values for: 
nginx.ingress.kubernetes.io/http2-push-preload 

nginx.ingress.kubernetes.io/ssl-passthrough 

hsts 

http2-max-requests 

keep-alive, it’s mentioned in the docs that this maps to functionality in http2 as well at the moment 

proxy-buffering 

proxy-stream-timeout 

ssl-redirect 

 

Tried setting annotation nginx.ingress.kubernetes.io/http2-push-preload: 'false' as described here https://github.com/rancher/rancher/issues/21324. 

Used an image with debug logs to check differences in headers for proxy request when rancher has the issue and when it doesn’t. Did not observe anything out of the ordinary. 
 

Asked in Kubernetes user slack in sig-cli channel. Some members believed it could be related to client used in kubectl. Originally thought it could be related to spdy issues but then realized the client for wait does not use spdy. Created an issue here https://github.com/kubernetes/kubernetes/issues/120282. 

Took stack traces for rancher and cli during the bug and posted kubectl one in issue. 

Increased verbosity of nic. Found nothing interesting. 

Tried different version of kubectl, did not resolve issue. 

Tried without webhook installed, did not resolve issue. 

Confirmed proper x-forwarded-for header is being used in proxy request. 

Flush in proxy client in rancher has a negative value which should cause it to flush imediately. Changed this value to 1 sec. This did not resolve the issue.

**Comment by moio** [comment]
We got a report that this is now fixed as of Rancher 2.10.1 and RKE2 v1.31.3+rke2r1.

So I am speculatively closing this issue, please reopen if symptoms are back at those versions or later.

Thanks everybody for the cooperation

**Comment by lalyos** [comment]
@moio can you please point to the PR/commit which supposed to fix this?
I just run into this issue, I have:
- rancher 2.12.3
- the imported cluster is k3s v1.34.1
- I use ingress nginx in front of rancher

I even tried the `use-http2: "false"`, it didn't help.
```
$ kubectl wait --for=condition=Available deploy/argocd-server -n argocd --timeout 0s
deployment.apps/argocd-server condition met
$ kubectl wait --for=condition=Available deploy/argocd-server -n argocd --timeout 1s
... hangs
```

**Comment by moio** [comment]
As mentioned, we got a report in which the original requestor had the problem fixed in a later version, but we are not sure what exactly fixed it.

I'm reopening the issue, although I am not sure when we will be able to look into it again. If you are a SUSE customer I recommend reaching out to Support for further assistance.
